### ISSUE TITLE
Make For-in (W089) suggest for-of with obj.keys() when esnext is true

### ISSUE THREAD

**Issue by McBrainy** [jshint/jshint#2976]
The below code gives "The body of a for in should be wrapped in an if statement to filter unwanted properties from the prototype.":

```javascript
// jshint esnext: true, forin: true
for (let x in obj) {
    console.log(x);
}
```

But the es2015 way of doing this is:

```javascript
for (let x of arr.keys()) {
    console.log(x);
}
```
My proposal is to return a message like: "For in may include unwanted properties from the prototype. Iterate on Object.prototype.keys() using for of instead."

**Comment by nicolo-ribaudo** [comment]
It is Object.keys(), not Object.prototype.keys()

**Comment by McBrainy** [comment]
@nicolo-ribaudo You're right; I was confusing it with `Array.prototype.keys()`. I just edited the issue.

**Comment by jugglinmike** [comment]
This could be helpful, but in truth, the warning you've mentioned is based on a very weak heuristic; users can still be caught off guard by the behavior of `for-in`. It's difficult to detect this problem statically.

In addition, I've have anecdotal evidence that developers are increasingly inclined to move away from the `for-in` statement altogether, both due to the tricky semantics and because it's generally harder to use. I think there's a valid argument to be made that in ES2015 code, the "for-in" statement is itself something of a hazard.

The feature request is essentially branching logic to make distinctions about the content of an existing warning. This might improve the experience somewhat, but it won't address the fundamental static analysis problem, and it doesn't serve the potential "for-in haters" I've described above.

In light of that, I'm thinking that we might actually serve more users with a new "enforcing" option: `noforin`. @McBrainy @rwaldron @nicolo-ribaudo Any thoughts on that?

**Comment by McBrainy** [comment]
I must be in the "for-in haters" camp, because I would probably use `noforin`. However, if the tool is going to make a suggestion for fixing a potential bug, why not make it the best suggestion available?
